clang/lib/3C/AVarBoundsInfo.cpp

```diff
@@ -1123,101 +1123,74 @@ BoundsKey AVarBoundsInfo::getCtxSensCEBoundsKey(const PersistentSourceLoc &PSL,
   return CSBKeyHandler.getCtxSensCEBoundsKey(PSL, BK);
 }
 
-void AVarBoundsInfo::computeArrPointers(ProgramInfo *PI) {
+void AVarBoundsInfo::computeArrPointers(const ProgramInfo *PI) {
   NtArrPointerBoundsKey.clear();
   ArrPointerBoundsKey.clear();
 
-  auto &CS = PI->getConstraints();
-  for (auto Bkey : PointerBoundsKey) {
-    // Regular variables.
-    auto &BkeyToPSL = DeclVarMap.right();
-    if (BkeyToPSL.find(Bkey) != BkeyToPSL.end()) {
-      auto &PSL = BkeyToPSL.at(Bkey);
-      if (hasArray(PI->getVarMap().at(PSL), CS)) {
-        ArrPointerBoundsKey.insert(Bkey);
-      }
-      // Does this array belong to a valid program variable?
-      if (isInSrcArray(PI->getVarMap().at(PSL), CS)) {
-        InProgramArrPtrBoundsKeys.insert(Bkey);
-      }
-
-      if (hasOnlyNtArray(PI->getVarMap().at(PSL), CS)) {
-        NtArrPointerBoundsKey.insert(Bkey);
-      }
-
-      continue;
+  // Called in following loop to add a BoundsKey to the appropriate sets based
+  // on the pointer type of a corresponding ConstraintVariable.
+  auto AddToArrSets = [this, PI](BoundsKey BK, const ConstraintVariable *CV) {
+    if (hasArray(CV, PI->getConstraints()))
+      ArrPointerBoundsKey.insert(BK);
+
+    // Does this array belong to a valid program variable?
+    if (isInSrcArray(CV, PI->getConstraints()))
+      InProgramArrPtrBoundsKeys.insert(BK);
+
+    if (hasOnlyNtArray(CV, PI->getConstraints())) {
+      NtArrPointerBoundsKey.insert(BK);
+      // If the return value is an nt array pointer and there are no declared
+      // bounds? Then, we cannot find bounds for this pointer.
+      // FIXME: This feels wrong. Why does it only apply to NTARR? Why does it
+      //        only apply to the return value? As I understand it, an
+      //        _Nt_array_ptr without bounds actually has bounds `count(0)`,
+      //        so it should be treated as a count bound for inference.
```

Comment:
Look in this: https://github.com/correctcomputation/checkedc-clang/pull/559 on why we added this for NTArr.